src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,99 @@
+// <copyright file="EventCountersMetrics.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Diagnostics.Tracing;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal class EventCountersMetrics : EventListener
+{
+    internal static readonly AssemblyName AssemblyName = typeof(EventCountersMetrics).Assembly.GetName();
+    internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentBag<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, Instrument> instruments = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
```

Comment:
What if folks changed the public part of `options` (e.g. `RefreshIntervalSecs`) after the `EventCountersMetrics` is instantiated?

Reply:
With how the interface is setup I don't think they really can. All they have access to is `AddEventCountersInstrumentation` and a callback function to make any changes to the `EventCountersInstrumentationOptions` defined in the `AddEventCountersInstrumentation` method.

Reply:
If this is no way that an end user could use `builder.AddInstrumentation(() => new something(options))`, then it's fine.

Btw, now I see that the ctor is for an internal class 👍

Reply:
I think they could change it, but I also think you'd be well within your rights to say that is undefined behavior if they do:
```C#
var globalOptions;
builder.AddEventCounterInstrumentation(opts =>
{
    globalOptions = opts;
}).Build();
globalOptions.RefreshIntervalSecs = 1_000;
```

If you want to be very defensive you can copy the options in the constructor so that you never again refer to the mutable options object that was exposed in the callback.